### ISSUE TITLE
Ambiguous signature for `extracting()` when using Kotlin

### ISSUE THREAD

**Issue by chrisly42** [assertj/assertj#1499]
#### Summary

When trying to use .extracting() with lambdas in kotlin, the call is ambigous and will not compile out-of-the-box. I think this is due to

```extracting(ThrowingExtractor<? super ELEMENT, V, EXCEPTION> extractor)```
and
```extracting(Function<? super ELEMENT, V> extractor)```

have the same effective signature because in kotlin, there are no checked exceptions (thank god).

#### Example

```kotlin
assertThat(highlights).hasSize(6).extracting(HighlightInfo::getText).doesNotContain("assertThat")
// nor does this work
assertThat(highlights).hasSize(6).extracting { it.text }.doesNotContain("assertThat")
```
I've got no idea on how to fix that in a backwards compatible manner other than to have a different method name for kotlin that won't clash or trying to add an extension that will cope with it.

**Comment by chrisly42** [comment]
Temporarily solved it by adding this kotlin extension and using the slightly (kotlin typical) misspelled version...

```
package org.assertj.core.api

import org.assertj.core.groups.FieldsOrPropertiesExtractor
import java.util.*

fun <SELF : AbstractIterableAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>,
        ACTUAL : Iterable<ELEMENT>,
        ELEMENT,
        ELEMENT_ASSERT : AbstractAssert<ELEMENT_ASSERT, ELEMENT>,
        V>
        AbstractIterableAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>.extrakting(extractor: (ELEMENT) -> V): AbstractListAssert<*, List<V>, V, ObjectAssert<V>> {
    val values = FieldsOrPropertiesExtractor.extract(actual, extractor)
    if (actual is SortedSet<*>) {
        usingDefaultElementComparator()
    }
    @Suppress("UNCHECKED_CAST")
    return newListAssertInstance(values).withAssertionState(myself) as AbstractListAssert<*, List<V>, V, ObjectAssert<V>>
}
```

**Comment by joel-costigliola** [comment]
I guess your solution is the way to go (or use assertk) as I'm not sure there is way of improving the Kotlin compatibility without changing the method names which is not not an option as it would be a breaking change.

**Comment by piefel** [comment]
Putting something in a different package feels a lot like a hack. I noticed that AssertJ 3.12 deprecated `Extractor` (in favour of `Function`), so there may still be a chance for this change in the future. How stable is assertk? I would not easily leave AssertJ just now.

**Comment by joel-costigliola** [comment]
`Extractor` is deprecated but `ThrowingExtractor` is not, the issue is not going away as (unfortunately) java has checked exceptions and this is what `ThrowingExtractor` is dealing with.

**Comment by piefel** [comment]
> […] I'm not sure there is way of improving the Kotlin compatibility without changing the method names which is not not an option as it would be a breaking change.

Perhaps a breaking change is needed from time to time. Method overloading that depends on whether or not exceptions are thrown is a bit weird anyway. How about `extractingWhileConvertingExceptions`?

**Comment by raharrison** [comment]
I've found that the below works for me:

`assertThat(highlights).hasSize(6).extracting<String> { it.text }.doesNotContain("assertThat")`

**Comment by DominiQN** [comment]
> I've found that the below works for me:
> 
> `assertThat(highlights).hasSize(6).extracting<String> { it.text }.doesNotContain("assertThat")`

but the below does not work...

```kotlin
assertThat(highlights)
    .hasSize(6)
    .extracting<String> { it.text }
    // still works
    .contains("this")
    // but from this point, 'Nothing' type is required
    .doesNotContain("that")
```

**Comment by joel-costigliola** [comment]
Closing since this is a Kotlin issue and there is not much we can do without breaking changes, moreover it seems there are some alternatives.

**Comment by TRUSTMEIMJEDI** [comment]
You can use it this way
```
assertThat(result).extracting<Tuple> {
    tuple(it.id, it.content, it.answer)
}.containsExactlyInAnyOrder(
    tuple(check, check1, check2),
    tuple(anotherCheck, anotherCheck1, anotherCheck2),
)
```

**Comment by scordio** [comment]
As Joel said, we're not that happy about making any breaking changes in version 3, but version 4 might be the right time to rethink how to better support Kotlin.

**Comment by ianbrandt** [comment]
I mostly use AssertJ with Kotlin these days. Even my teammates that still code mostly in Java are writing their tests in Kotlin. AssertJ generally works great from Kotlin, but little things like this would make it even better. I'm psyched to see this re-opended, and hear of the potential for even better Kotlin compatibility in 4.0.

**Comment by bric3** [comment]
By the way it seems that `flatExtracting` is also exposed to this issue.

```
assertThat(notificationTestUtil.listOurNotifications())
            .flatExtracting(Notification::getTitle, Notification::getContent)
            .containsOnly("No 'origin' provided")
```

=> `[NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER] Not enough information to infer type variable EXCEPTION`

**Comment by realdadfish** [comment]
I agree with others that a small shim library on top of assertj-core might already help a lot, similar to what Google did with the *-ktx libraries, so people could use the more idiomatic Kotlin versions of things, all under the assertj umbrella, so we don't have to duplicate little things like this again and again and again:

```kotlin
inline fun <reified T : Any> ObjectAssert<*>.asInstanceOf(): ObjectAssert<T> =
        asInstanceOf(InstanceOfAssertFactories.type(T::class.java))
```

Would be willing to contribute if there is interest.

**Comment by scordio** [comment]
Would it be really necessary to have a separate library on top of `assertj-core` or could we do something like JUnit did in [`Assertions.kt`](https://github.com/junit-team/junit5/blob/main/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt)?

**Comment by bric3** [comment]
Having a separate kotlin library may impede discovery of such library. The JUnit approach is in my opinion a sensible approach, yet I wonder about the actual feasibility or consistency of the API in this case.

**Comment by scordio** [comment]
Likely, both `AbstractObjectArrayAssert::extracting` and `AbstractIterableAssert::extracting` are affected.

We should add corresponding test cases in the `assertj-core-kotlin` test module.

**Comment by scordio** [comment]
Based on the analysis done today with @patrickwalter214 at [Hack Commit Push Winterthur](https://winterthur2025.hack-commit-pu.sh/), we identified two possible changes to enhance Kotlin compatibility without major breaks of Java use cases:
* Remove `extracting(Function)` variants, as `extracting(ThrowingExtractor)` can cover them too
* Remove the `EXCEPTION` type parameter of `ThrowingExtractor`
  * Originally requested at https://github.com/assertj/assertj/issues/755#issuecomment-261728955 and introduced in #963
  * Further analized at https://github.com/assertj/assertj/issues/755#issuecomment-2970370829

In addition, we might need to check:
* `flatExtracting(Function)` vs. `flatExtracting(ThrowingExtractor)`

**Comment by PatrickWalter214** [comment]
I opened a PR for the issue reported originally, altough while working on the issue I discovered the following additional issues:

- The same change is most likely also necessary for the, already mentioned by @scordio, `flatExtracting` method, but also for the `map` and `flatMap` methods when asserting an `Iterable` with `AbstractIterableAssert`.
- The same change is also necessary for equivalent methods when asserting arrays with `AbstractObjectArrayAssert` and `AtomicReferenceArrayAssert`.
- The issue reported by @DominiQN that in Kotlin chaining `contains(...).doesNotContain(...)` does not work still exists.

I can open additional issues for the mentioned problems, verifying before that the issue really affects the `map`/`extracting` methods and their `flat..` equivalents. How many issues should I theoretically open, one by method, one by class, one single issue, or do we just keep this issue open?

**Comment by scordio** [comment]
> Remove `extracting(Function)` variants, as `extracting(ThrowingExtractor)` can cover them too

This decision turned out to be not a good one, from a compatibility perspective, and will be reconsidered at https://github.com/assertj/assertj/pull/3854#discussion_r2549110660.

**Comment by scordio** [comment]
Since we're currently working on version 3.28, we can offer a more gentle migration path instead of removing the `extracting(ThrowingExtractor)` methods in version 4 without notice.
